### ISSUE TITLE
How to document the `[]` (postfix) operator

### ISSUE THREAD

**Issue by judofyr** [sanity-io/GROQ#2]
## What should it be called?

- "Array traversal operator": This is what we've always(?) called it
- "Projection operator": I thought about this earlier and it makes some sense. It's very similar to a projection (in the sense that it changes a stream of data); the only difference is that the output can be *any* value (not only an object).

## How do we specify the semantics

There are three different ways as far as I can see:

### 1. Syntactical transformation semantics

In the grammar the operator is a regular postfix operator, and we introduce two new functions (`map` and `flatten`) and specify the operator in terms of a syntactical transformation which happens before processing:

- `foo[].bar[0]` => `foo | map(bar[0])`
- `(foo[].bar)[0]` => `(foo | map(bar))[0]`
- `foo[].bar[].baz` => `foo | map(bar) | flatten() | map(baz)`

Pros/cons:

- We need to introduce two new functions.
- The grammar is trivial: it's a "regular" postfix operator.
- The `LHS.foo`-expression is the same type regardless of whether the LHS is a `[]`-expression or a regular expression. (Since it's dissolved after the processing.)
- We don't need to say anything specific in the filter/projection/slice/etc. expression spec text.

### 2. Grammar-level semantics

We enforce it inside the grammar:

```
plainexpr
  := ident
   | "@" | "*" | …

compoundexpr
  := compoundexpr "[" EXPR "]"
   | compoundexpr "." IDENT
   | plainexpr

arrexpr
  := compoundexpr "[]"
   | arrexpr "[]"
   | arrexpr "[" EXPR "]"
   | arrexpr "." IDENT

expr := compoundexpr | arrexpr

Pros/cons:

- The grammar needs to duplicate filter/slice/attribute/etc. expressions.
- We end up with duplicate syntactical nodes for array filter vs. regular filter. These will be duplicated in the spec.
- I'm not 100% sure how to structure the grammar to avoid ambiguities.
- The exact semantics are more precise (a strict EBNF grammar is far more precise than any specification text we write).
```

### 3. Value-level semantics

In the grammar the operator is just a regular postfix operator. Evaluating that operator returns a `Mapper`-value. The `Mapper` type is a subtype of `Array`. The specification of filter expression would then have a part like this:

> If the base value (e.g. left-hand side) is a Mapper value, then it creates a new Mapper with [same semantics as previously].

Pros/cons:

- We need to incorporate this into the filter/slice/attribute/etc. expression spec texts.
- The grammar is trivial: it's a "regular" postfix operator.
- It's probably not how you'd want to implement this operator in practice.
- It's a quite naïve way (i.e: might be intuitive)

**Comment by judofyr** [comment]
I'm principally not a fan of (2) because I think pushing semantics down into the grammar often complicates things too much.

I'm slightly preferring (3) because I think it's easier to explain runtime semantics than it is to explain a compilation-time syntactical transformation.

**Comment by simen** [comment]
Are we effed if we in some version of the future create a REPL, and then someone does a `f := foo[]` and expect `f` to be a Mapper? Maybe it would be a good idea to avoid variables in our hypothtical REPL anyway, but the point stands. Are we creating an abstraction monster that will come back to haunt us if we do (2)?

**Comment by judofyr** [comment]
> Are we effed if we in some version of the future create a REPL, and then someone does a `f := foo[]` and expect `f` to be a Mapper?

Depends on what the semantics are:

```
// Is this equivalent to:
f := foo[]
f.bar

// (a)
(foo[]).bar

// (b)
foo[].bar
```

Alternative (b) essentially forces solution (3) (and then we need to expose a presentation of the Mapper-type in the REPL). If we go for alternative (a) then all three works, and in the case of solution (3) the Mapper-type is being realized into a plain Array (similarly to how parenthesis works now).

**Comment by atombender** [comment]
> "Array traversal operator": This is what we've always(?) called it

I don't love it, but I think that's the best we've got. Another option is "array mapping operator".

jq calls it an array construction operator, but it has different rules (you can do `a.[].b`, and `[1, 2, 3]` is syntactic sugar for applying the `[]` constructor to three numbers) and is certainly different from GROQ in many weird ways.

> Are we effed if we in some version of the future create a REPL, and then someone does a f := foo[] and expect f to be a Mapper?

I don't like that. Especially as we already have the syntax rule that `(a[]).b` is the same as `a.b` (i.e. parens abort mapping).

**Comment by atombender** [comment]
> 1. Syntactical transformation semantics
> ...
> foo[].bar[0] => foo | map(bar[0])

Minor point: The `|` operator works on both values and arrays, so this is not strictly the right transformation. The expression must terminate if `foo` isn't an array. The transformation requires some kind of assertion to be correct, like:

```
foo[].bar[0] => assertArray(foo) | map(bar[0])
```

(`assertArray()` would presumably return null if not an array.)

We don't have lambda expressions, so `map(bar[0])` cannot be done cleanly in GROQ at this point. One solution is to let functions take expressions, and define the signature for `map` as `map(expression)`. We have a similar problem with `order()` where we totally cheat.

**Comment by judofyr** [comment]
> Minor point: The `|` operator works on both values and arrays, so this is not strictly the right transformation. The expression must terminate if `foo` isn't an array.

The expression doesn't necessarily has to terminate as long as it turns `null` and accessing/working on null always returns null. 

> We don't have lambda expressions, so `map(bar[0])` cannot be done cleanly in GROQ at this point. One solution is to let functions take expressions, and define the signature for `map` as `map(expression)`. We have a similar problem with `order()` where we totally cheat.

I think `map(expression)` is perfectly fine. I understand that it can be strange that function arguments are sometimes eagerly and sometimes lazily + scope-inducing, but it seems like an intuitive way to provide lambdas without having an explicit syntax for it.

**Comment by atombender** [comment]
> The expression doesn't necessarily has to terminate as long as it turns null and accessing/working on null always returns null.

Yes, but my point is that you can't rewrite it exactly like that. They're not equivalent; assuming `foo` is `{"bar": 123"}`, then:

```
foo[]
  => null
foo[].bar
  => null
foo | bar
  => 123
foo[] | bar
  => null (not sure if we have a test case for this!)
```

> I think map(expression) is perfectly fine. I understand that it can be strange that function arguments are sometimes eagerly and sometimes lazily + scope-inducing, but it seems like an intuitive way to provide lambdas without having an explicit syntax for it.

I don't think it's particularly intuitive, but it must also be expressed at the API (documentation) level.

Another point is that `|` itself is our mapping operator:

```
{"a": 42} | a
  => 42
[{"a": 42}] | a
  => [42]
[{"a": 42, "b": 43}] | ([a, b])
  => [42, 43]
```

We had a discussion some time ago what the semantics of `|` should be and how to explain it. One point of contention is that it currently doesn't just map one-by-one, but aggregates, e.g. in `* | order(a)`, this is like handing the input as an argument: `order(*, a)`. Erik argued that such functions (think `count`, `min`, `sum`, `groupBy`, etc.) could be expressed in terms of taking a "stream" as its first argument.

**Comment by judofyr** [comment]
> Yes, but my point is that you can't rewrite it exactly like that. They're not equivalent; assuming foo is `{"bar": 123"}`, then:

I'm not talking about rewriting it to any existing expression/syntax. I'm talking about introducing a complete new function `map` which has the exact same semantics as array traversal today. In a similar fashion we *could* introduce a `filter`-function such that `*[_type == "bar"]` is "just" syntax sugar for `* | filter(_type == "bar")`.

---

The rest here is far away from the issue at hand, but here it goes anyway:

> Another point is that `|` itself is our mapping operator:

Is it? The reference docs on sanity.io says that the RHS of a pipe can be a parenthesis-expression, but there's no mention that `foo | bar` is valid syntax. Is there a reason why we should allow it? It adds another level ambiguity.

> We had a discussion some time ago what the semantics of `|` should be and how to explain it.

The way I see it there isn't a pipe operator at all; it's just a syntactical token which is used in some expressions:

- Filter/slice/access expression: `expr '|'? '[' expr ']'`
- Projection expression: `expr  '|'? '{' … '}'`
- Map expression: `expr '|' '(' expr ')'`
- Call expression: `expr '|' ident '(' … ')'`

I don't see any reason in trying to explain `|` separately (just like we don't have a section explaining all places where you use a `,`).

(I also wish that `|` was never optional (e.g. we only supported either `foo{bar}` xor `foo|{bar}`), but I think it's too late for that.)
